l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,444 @@
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner whose attester public key will be changed.
+    /// @param _pubKey The new ECDSA public key to assign to the node's attester.
+    function changeAttesterKey(
+        address _nodeOwner,
+        Secp256k1PublicKey calldata _pubKey
+    ) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        _verifyInputSecp256k1PublicKey(_pubKey);
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
+        if (_finalizeNodeRemoval(_nodeOwner, node)) {
+            return;
+        }
+
+        bytes32 prevHash = _hashAttesterPubKey(node.attesterLatest.pubKey);
+        delete attesterPubKeyHashes[prevHash];
+        bytes32 newHash = _hashAttesterPubKey(_pubKey);
+        attesterPubKeyHashes[newHash] = true;
+
+        _ensureAttesterSnapshot(node);
+        node.attesterLatest.pubKey = _pubKey;
+
+        emit NodeAttesterKeyChanged(_nodeOwner, _pubKey);
+    }
+
+    /// @notice Adds a new commit to the validator committee.
+    /// @dev Implicitly updates the validator committee by affecting off-chain readers based on the current state of a node's validator attributes:
+    /// - If "validatorsCommit" > "node.validatorLastUpdateCommit", read "node.validatorLatest".
+    /// - If "validatorsCommit" == "node.validatorLastUpdateCommit", read "node.validatorSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitValidators() external onlyOwner {
+        validatorsCommit++;
+
+        emit ValidatorsCommitted();
+    }
+
+    /// @notice Adds a new commit to the attester committee.
+    /// @dev Implicitly updates the attester committee by affecting off-chain readers based on the current state of a node's attester attributes:
+    /// - If "attestersCommit" > "node.attesterLastUpdateCommit", read "node.attesterLatest".
+    /// - If "attestersCommit" == "node.attesterLastUpdateCommit", read "node.attesterSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitAttesters() external onlyOwner {
+        attestersCommit++;
+
+        emit AttestersCommitted();
+    }
+
+    function numNodes() public view returns (uint256) {
+        return nodeOwners.length;
+    }
+
+    function _finalizeNodeRemoval(address _nodeOwner, Node storage _node) private returns (bool) {
+        if (
+            _node.attesterSnapshot.pendingRemoval &&
+            _node.validatorSnapshot.pendingRemoval
+        ) {
+            nodeOwners[_nodeOwnerIdx(_nodeOwner)] = nodeOwners[nodeOwners.length - 1];
+            nodeOwners.pop();
+            delete nodes[_nodeOwner];
+
+            delete attesterPubKeyHashes[_hashAttesterPubKey(_node.attesterLatest.pubKey)];
+            delete validatorPubKeyHashes[_hashValidatorPubKey(_node.validatorLatest.pubKey)];
+
+            return true;
```

Comment:
Should this perhaps also emit some event to show this happened? I think it might be useful because the user could call for example `changeAttesterKey` and if the node doesn't exist they get a revert, if all is well they get `NodeAttesterKeyChanged`, but if the record existed but was pending removal they can only figure out that no update happened by the _absence_ of the expected event. Since everything else is so explicit, this could have its signal as well.

Reply:
Agree, added here: https://github.com/matter-labs/era-contracts/pull/555/commits/7a2a755d490cf15bbc3048722dd51c6be59de5c5